### PR TITLE
Bugfix/abi encode

### PR DESCRIPTION
**Pull request by uv-orbs**

### CODE REVIEW

**Comment by uv-orbs** [comment]
name is not good, it actually fixed lots of other stuff

File: models/abi_order.go [hunk 1]
```diff
@@ -39,6 +39,42 @@ type AbiFragment struct {
 	Outputs                []PartialOutput
 }
 
+// Custom JSON serialization for *big.Int fields
```
Comment (Luke-Rogerson) [inline models/abi_order.go:42]: Comments can be removed

File: service/amount_out.go [hunk 2]
```diff
@@ -57,6 +57,10 @@ func getOutAmountInAToken(ctx context.Context, it models.OrderIter, inAmountB de
 	var order *models.Order
 	for it.HasNext() && inAmountB.IsPositive() {
 		order = it.Next(ctx)
+		if order == nil {
+			logctx.Error(ctx, "order::it.Next() returned nil")
```
Comment (Luke-Rogerson) [inline service/amount_out.go:61]: Can you update the Error log to Warn then as this is not an unexpected error, and change the error? You could do models.ErrInsufficientLiquidity for example?

File: service/amount_out.go [hunk 3]
```diff
@@ -95,6 +99,10 @@ func getOutAmountInBToken(ctx context.Context, it models.OrderIter, inAmountA de
 	var frags []models.OrderFrag
 	for it.HasNext() && inAmountA.IsPositive() {
 		order = it.Next(ctx)
+		if order == nil {
```
Comment (Luke-Rogerson) [inline service/amount_out.go:102]: Can you update the Error log to Warn then as this is not an unexpected error, and change the error? You could do models.ErrInsufficientLiquidity for example?

File: transport/rest/taker.go [hunk 4]
```diff
@@ -93,19 +93,23 @@ func (h *Handler) nameFromAddress(name, address string) (string, error) {
...
-	OutName, err := h.nameFromAddress(req.OutToken, req.OutTokenAddress)
-	if err != nil {
-		return err
+	if req.InToken == "" {
+		InName, err := h.nameFromAddress(req.InToken, req.InTokenAddress)
+		if err != nil {
+			return err
+		}
+		if len(InName) > 0 {
+			req.InToken = InName
```
Comment (Luke-Rogerson) [inline transport/rest/taker.go:102]: So `InName` is a slice / array ? Isn't this assingment wrong then? You want InToken to be `BSC`, not `['BSC']`.
Also, is it possible that `InName` will return more than 1 elements?

File: transport/rest/taker.go [hunk 5]
```diff
@@ -93,19 +93,23 @@ func (h *Handler) nameFromAddress(name, address string) (string, error) {
...
+		}
 	}
-	if len(OutName) > 0 {
-		req.OutToken = OutName
+	if req.OutToken == "" {
+		OutName, err := h.nameFromAddress(req.OutToken, req.OutTokenAddress)
+		if err != nil {
+			return err
+		}
+		if len(OutName) > 0 {
```
Comment (Luke-Rogerson) [inline transport/rest/taker.go:110]: Same comment as above

File: service/amount_out.go (hunk 2 above)
Comment (uv-orbs) [reply service/amount_out.go:61]: i handled it in the calling func to return insufficient liqidity
`it = s.orderBookStore.GetMaxBid(ctx, symbol)
		if it == nil {
			logctx.Error(ctx, "GetMaxBid failed no orders in iterator")
			return models.QuoteRes{}, models.ErrIterFail
		}
		if !it.HasNext() {
			logctx.Warn(ctx, "GetMaxBid failed no orders in iterator")
			return models.QuoteRes{}, models.ErrInsufficientLiquity
		}`

Comment (uv-orbs) [reply service/amount_out.go:61]: so i leave this code as it was, since it is indeed not expected

File: transport/rest/taker.go (hunk 4 above)
Comment (uv-orbs) [reply transport/rest/taker.go:102]: mmm nope. inName is a string, not a slice. internally we work with stokenNames, but deddy may call us with either a name or an addrsss. if he comes with and adress and the inToken name eq "" empty string, we resolve that, hence the assignment

**Pull request merged**